src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -0,0 +1,134 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "frozenobjectheap.h"
+#include "memorypool.h"
+
+FrozenObjectHeap::FrozenObjectHeap():
+    m_pStart(nullptr),
+    m_pCurrent(nullptr),
+    m_SizeCommited(0),
+    m_Size(0),
+    m_SegmentHandle(nullptr)
+    COMMA_INDEBUG(m_ObjectsCount(0))
+{
+    m_PageSize = GetOsPageSize();
+    m_Crst.Init(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC);
+}
+
+#define FOH_RESERVE_PAGES 1024 // e.g. reserve 4Mb of virtual memory
+#define FOH_COMMIT_PAGES 32  // e.g. commit 128Kb chunks
+
+FrozenObjectHeap::~FrozenObjectHeap()
+{
+    if (m_SegmentHandle != nullptr)
+    {
+        GCHeapUtilities::GetGCHeap()->UnregisterFrozenSegment(m_SegmentHandle);
+    }
+
+    if (m_pStart != nullptr)
+    {
+        ClrVirtualFree(m_pStart, 0, MEM_RELEASE);
+    }
+}
+
+bool FrozenObjectHeap::Initialize()
+{
+    m_Size = m_PageSize * FOH_RESERVE_PAGES;
+
+    _ASSERT(FOH_RESERVE_PAGES > FOH_COMMIT_PAGES);
+    _ASSERT(FOH_RESERVE_PAGES % FOH_COMMIT_PAGES == 0);
+    _ASSERT(m_PageSize > MIN_OBJECT_SIZE);
+    _ASSERT(m_SegmentHandle == nullptr);
+    _ASSERT(m_pStart == nullptr);
+
+    void* alloc = ClrVirtualAllocAligned(nullptr, m_Size, MEM_RESERVE, PAGE_READWRITE, m_PageSize);
+    if (alloc != nullptr)
+    {
+        // Commit FOH_COMMIT_PAGES pages in advance
+        alloc = ClrVirtualAllocAligned(alloc, m_PageSize * FOH_COMMIT_PAGES, MEM_COMMIT, PAGE_READWRITE, m_PageSize);
+    }
+
+    if (alloc != nullptr)
```

Comment:
We should throw exception instead of failing silently.

Reply:
Currently it's just an optimization and if COMMIT fails (from my understanding it might happen during hard memory pressure on OS) it just asks caller to try other heaps 🙂

Reply:
And the commit fails in other heaps. I do not see the point in bouncing the problem around.